### PR TITLE
Serving input payload doc

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/12848?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12848/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12848
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Add some explanation between two saved files when using input_example

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: docs/source/_static/images/models/serving_input_payload_json.png
Comment (harupy) [inline docs/source/_static/images/models/serving_input_payload_json.png:1]: Do we need images? Can we use code blocks?

Comment (serena-ruan) [reply docs/source/_static/images/models/serving_input_payload_json.png:1]: Is it because it's not pretty?

Comment (harupy) [reply docs/source/_static/images/models/serving_input_payload_json.png:1]: Several reasons:

1. They increase the repo size = slower clone = slow development iterations
2. More efforts are required to update them.
3. They become outdated pretty quick.

Comment (harupy) [reply docs/source/_static/images/models/serving_input_payload_json.png:1]: If the goal is just to show _**what the examples look like**_, I think code blocks are sufficient.

Comment (serena-ruan) [reply docs/source/_static/images/models/serving_input_payload_json.png:1]: Sounds good, let me update to a table

Comment (harupy) [reply docs/source/_static/images/models/serving_input_payload_json.png:1]: Thanks! Adding two images don't make much difference, but we blindly add many images, we'll regret.

Comment (serena-ruan) [reply docs/source/_static/images/models/serving_input_payload_json.png:1]: Updated 
image

File: docs/source/model/signatures.rst [hunk 1]
```diff
@@ -748,7 +748,51 @@ model signatures in log_model calls when signatures aren't specified.
...
+            python_model=MyModel(),
+            artifact_path="model",
+            input_example={"question": "What is MLflow?"},
+        )
+
+Example files:
+
+ +---------------------------------------+-----------------------------------------+
+ |  input_example.json                   |  serving_input_payload.json             |
+ +---------------------------------------+-----------------------------------------+
```
Comment (harupy) [inline docs/source/model/signatures.rst:782]: Not a blocker, but I'd vertically stack the examples for editability

```
.. code-block::
   
     # input_example.json

.. code-block::
   
     # serving_input_payload.json
```

File: docs/source/models.rst [hunk 2]
```diff
@@ -125,6 +125,7 @@ In MLflow, understanding the intricacies of model signatures and input examples
 
 - **Model Signature**: Defines the schema for model inputs, outputs, and additional inference parameters, promoting a standardized interface for model interaction.
 - **Model Input Example**: Provides a concrete instance of valid model input, aiding in understanding and testing model requirements. Additionally, if an input example is provided when logging a model, a model signature will be automatically inferred and stored if not explicitly provided.
+- **Model Serving Payload Example**: Provides a json payload example for querying a deployed model endpoint. If an input example is provided when logging a model, a serving paylod example is automatically generated from input example and saved into ``serving_input_payload.json`` file.
```
Comment (harupy) [inline docs/source/models.rst:128]:
```suggestion
- **Model Serving Payload Example**: Provides a json payload example for querying a deployed model endpoint. If an input example is provided when logging a model, a serving paylod example is automatically generated from the input example and saved as ``serving_input_payload.json``.
```

File: docs/source/model/signatures.rst [hunk 3]
```diff
@@ -748,7 +748,51 @@ model signatures in log_model calls when signatures aren't specified.
     Secondly, and just as importantly, it **validates the model's requirements**. This input example is utilized to execute a prediction 
     using the model that is about to be logged, thereby enhancing the accuracy in identifying model requirement dependencies.
     It is **highly recommended** to always include an input example along with your models when you log them.
+    Since MLflow 2.16.0, if the provided input example fails on validation, a warning is issued with an instruction to validate the input example
+    prior to model deployment.
 
+Since MLflow 2.16.0, when logging a model with an input example, there are two files saved into the model's artifacts directory:
+
+- ``input_example.json``: The input example in JSON format.
+- ``serving_input_payload.json``: The input example converted to JSON format for querying a deployed model endpoint.
```
Comment (B-Step62) [inline docs/source/model/signatures.rst:757]:
```suggestion
- ``serving_input_payload.json``: The input example in JSON format, with additional transformation to have compatible schema for querying a deployed model REST endpoint.
```
nit: To make the difference even clearer...

File: docs/source/model/signatures.rst (hunk 1 above)
Comment (B-Step62) [reply docs/source/model/signatures.rst:782]: I think we can use [list-table](https://github.com/mlflow/mlflow/blob/d828b72897415548601e9825c43485e13136c96f/docs/source/llms/transformers/task.rst?plain=1#L77), much easier to edit imo

File: docs/source/model/signatures.rst [hunk 4]
```diff
@@ -748,7 +748,51 @@ model signatures in log_model calls when signatures aren't specified.
...
+ |        "question": "What is MLflow?"  |       "inputs": {                       |
+ |    }                                  |         "question": "What is MLflow?"   |
+ |                                       |       }                                 |
+ |                                       |   }                                     |
+ |                                       |                                         |
+ +---------------------------------------+-----------------------------------------+
+
+
+``input_example.json`` contains the input example in its original format, while ``serving_input_payload.json`` is a JSON-serialized version of the input example with
+a defined ``key`` (one of ``dataframe_split, instances, inputs or dataframe_records``) that mlflow scoring server requires when `querying a deployed model endpoint <../../cli.html#mlflow-models-serve>`_.
```
Comment (B-Step62) [inline docs/source/model/signatures.rst:795]:
```suggestion
one of predefined keys (``"dataframe_split"``, ``"instances"``, ``"inputs"``, and ``"dataframe_records"``) that mlflow scoring server requires when `querying a deployed model endpoint <../../cli.html#mlflow-models-serve>`_.
```
super-nit

File: docs/source/model/signatures.rst [hunk 5]
```diff
@@ -748,7 +748,51 @@ model signatures in log_model calls when signatures aren't specified.
     Secondly, and just as importantly, it **validates the model's requirements**. This input example is utilized to execute a prediction 
     using the model that is about to be logged, thereby enhancing the accuracy in identifying model requirement dependencies.
     It is **highly recommended** to always include an input example along with your models when you log them.
+    Since MLflow 2.16.0, if the provided input example fails on validation, a warning is issued with an instruction to validate the input example
```
Comment (B-Step62) [inline docs/source/model/signatures.rst:751]: nit: not sure if we have to give a notice of "we will emit warning" here, as the warning itself is pretty self-explanatory...?

File: docs/source/model/signatures.rst [hunk 6]
```diff
@@ -748,7 +748,51 @@ model signatures in log_model calls when signatures aren't specified.
...
+
+
+    with mlflow.start_run():
+        model_info = mlflow.pyfunc.log_model(
+            python_model=MyModel(),
+            artifact_path="model",
+            input_example={"question": "What is MLflow?"},
+        )
+
+Example files:
```
Comment (B-Step62) [inline docs/source/model/signatures.rst:778]:
```suggestion
Example files created by MLflow:
```

File: docs/source/model/signatures.rst (hunk 4 above)
Comment (B-Step62) [reply docs/source/model/signatures.rst:795]: Btw the link for `cli.html` is broken. Can we link to https://mlflow.org/docs/latest/deployment/deploy-model-locally.html#local-inference-server-spec instead, which has more comprehensive guidance?

File: docs/source/model/signatures.rst [hunk 7]
```diff
@@ -748,7 +748,51 @@ model signatures in log_model calls when signatures aren't specified.
...
+ |    {                                  |   {                                     |
+ |        "question": "What is MLflow?"  |       "inputs": {                       |
+ |    }                                  |         "question": "What is MLflow?"   |
+ |                                       |       }                                 |
+ |                                       |   }                                     |
+ |                                       |                                         |
+ +---------------------------------------+-----------------------------------------+
+
+
+``input_example.json`` contains the input example in its original format, while ``serving_input_payload.json`` is a JSON-serialized version of the input example with
```
Comment (B-Step62) [inline docs/source/model/signatures.rst:794]: nit: Maybe this description can go inside the table row.

**Review by B-Step62 — APPROVED** [review]
LGTM once a few cosmetic comments are addressed! I like the idea of having two JSON files side-by-side:)

File: docs/source/model/signatures.rst (hunk 1 above)
Comment (serena-ruan) [reply docs/source/model/signatures.rst]: Updated 
image

**Review by harupy — APPROVED** [review]
LGTM

File: docs/source/model/signatures.rst [hunk 8]
```diff
@@ -749,12 +749,65 @@ model signatures in log_model calls when signatures aren't specified.
...
+     - The input example in its original format.
+   * - serving_input_payload.json
+     - 
+        .. code-block:: json
+
+                {
+                    "inputs": {
+                        "question": "What is MLflow?"
+                    }
+                }
```
Comment (harupy) [inline docs/source/model/signatures.rst:802]:
```suggestion
        .. code-block:: json

            {
              "inputs": {
                "question": "What is MLflow?"
              }
            }
```

File: docs/source/model/signatures.rst [hunk 9]
```diff
@@ -749,12 +749,65 @@ model signatures in log_model calls when signatures aren't specified.
...
+   * - File name
+     - Content
+     - Explanation
+   * - input_example.json
+     - 
+        .. code-block:: json
+
+            {
+                "question": "What is MLflow?"
+            }
```
Comment (harupy) [inline docs/source/model/signatures.rst:791]:
```suggestion
            {
              "question": "What is MLflow?"
            }
```

**Pull request merged**